components/authn-service/tokens/pg/adapter.go

```diff
@@ -74,38 +87,76 @@ func (a *adapter) insertToken(ctx context.Context,
 }
 
 func (a *adapter) UpdateToken(ctx context.Context,
-	id, description string, active bool, projects []string) (*tokens.Token, error) {
+	id, description string, active bool, updatedProjects []string) (*tokens.Token, error) {
+
+	ctx, cancel := context.WithCancel(ctx)
+	defer cancel()
+
+	tx, err := a.db.BeginTx(ctx, nil)
+	if err != nil {
+		return nil, processSQLError(err, "get projects transaction")
+	}
+
 	projectsFilter, err := ProjectsListFromContext(ctx)
 	if err != nil {
 		return nil, processSQLError(err, "get projects filter for tokens")
 	}
 	t := tokens.Token{}
 
 	// ensure we do not pass null projects to db
-	if projects == nil {
-		projects = []string{}
+	if updatedProjects == nil {
+		updatedProjects = []string{}
 	}
+
+	var originalProjects []string
 	var row *sql.Row
+	err = tx.QueryRowContext(ctx,
+		`SELECT project_ids FROM chef_authn_tokens
+		WHERE id=$1 AND projects_match(project_ids, $5::TEXT[])
+		FOR UPDATE;`,
+		id, pq.Array(projectsFilter)).Scan(pq.Array(&originalProjects))
+	if err != nil {
+		return nil, processSQLError(err, "fetch projects for update")
+	}
+
+	projectDiff := projectassignment.CalculateProjectDiff(originalProjects, updatedProjects)
+	if len(projectDiff) != 0 {
+		_, err := a.authzClient.ValidateProjectAssignment(ctx, &authz_v2.ValidateProjectAssignmentReq{
+			Subjects:   auth_context.FromContext(auth_context.FromIncomingMetadata(ctx)).Subjects,
+			ProjectIds: projectDiff,
+		})
+		if err != nil {
+			// TODO verify/decide if this is what we want
+			// return error unaltered because it's already a GRPC status code
+			return nil, err
+		}
+	}
+
+	// TODO why are we doing this? What if they want to delete the description?
 	if description != "" {
 		row = a.db.QueryRowContext(ctx,
```

Comment:
Since we're opening a tx above, we should probably use it here. Otherwise, this request will be _outside_ of the transaction.